scalyr_agent/log_processing.py

```diff
@@ -88,45 +88,150 @@
 
 log = scalyr_logging.getLogger(__name__)
 
+class CRIParseError(Exception):
+    def __init__(self, line, message):
+        self.line = line
+        self.message = message
 
-def _parse_cri_log(line):
-    """
-    Parse a log line that uses the K8S Container Runtime Interface (CRI) format
-    https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/kubelet-cri-logging.md
-    """
-    # find the end of the timestamp segment
-    index = line.find(" ")
-    if index < 0:
-        return None, None, None, None
+    def __repr__(self):
+        return "Error parsing line - %s: %s" % self.message, self.line
+
+class CRILogLine(object):
+    __slots__ = ("timestamp", "stream", "tags", "message", "raw_timestamp", "is_partial", "is_full")
+
+    def __init__(self, timestamp, stream, tags, message, raw_timestamp):
+        self.timestamp = timestamp
+        self.stream = stream
+        self.tags = tags
+        self.message = message
+        self.raw_timestamp = raw_timestamp
+
+        self.is_partial = "P" in tags if tags else False
+        self.is_full = "F" in tags if tags else False
+
+    @classmethod
+    def from_raw_line(cls, raw_line):
+        decoded_raw_line = raw_line.decode("utf-8", "replace")
+        timestamp, stream, tags, message, raw_timestamp = cls.__parse_cri_log(
+            decoded_raw_line
+        )
+
+        log_line = cls(timestamp, stream, tags, message, raw_timestamp)
+
+        if not log_line.is_partial and not log_line.is_full:
+            raise CRIParseError("Expected tags to contain P or F", raw_line)
+
+        return log_line
 
-    # parse the timestamp
-    timestamp = rfc3339_to_nanoseconds_since_epoch(line[:index])
-    if timestamp is None:
-        return None, None, None, None
+    @staticmethod
+    def __parse_cri_log(line):
+        """
+        Parse a log line that uses the K8S Container Runtime Interface (CRI) format
+        https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/kubelet-cri-logging.md
+        """
+        parts = line.split(" ", maxsplit=3)
+        # The 4 parts are the timestamp, stream, tags, and message. The 4th part can contain spaces.
+        if len(parts) < 3:
+            raise CRIParseError("Expected 3 or 4 records (timestamp, stream, tags, optional message), got %d" % len(parts), line)
 
-    # skip over the ' ' delimiter and parse the stream type
-    line = line[index + 1 :]
-    index = line.find(" ")
-    if index < 0:
-        return None, None, None, None
+        # parse the timestamp
+        timestamp = rfc3339_to_nanoseconds_since_epoch(parts[0])
+        if timestamp is None:
+            raise CRIParseError("Could not parse timestamp", line)
+        raw_timestamp = timestamp
```

Comment:
A minor thing, it might be a little clearer to rename raw_timestamp to unixtime or epochtime